sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -288,6 +316,19 @@ public ProcessContinuation processElement(
               Optional.ofNullable(watermarkEstimator.currentWatermark()));
     }
     try (Consumer<byte[], byte[]> consumer = consumerFactoryFn.apply(updatedConsumerConfig)) {
+      // Check whether current TopicPartition is still available to read.
+      Set<TopicPartition> existingTopicPartitions = new HashSet<>();
+      for (List<PartitionInfo> topicPartitionList : consumer.listTopics().values()) {
+        topicPartitionList.forEach(
+            partitionInfo -> {
+              existingTopicPartitions.add(
+                  new TopicPartition(partitionInfo.topic(), partitionInfo.partition()));
+            });
+      }
+      if (!existingTopicPartitions.contains(kafkaSourceDescriptor.getTopicPartition())) {
+        return ProcessContinuation.stop();
```

Comment:
What's about checkpointing for already read records in this case? Are they going to be processed?

Reply:
I'm assuming that you are asking about the resumed `KafkaSourceDescriptor`. It depends on the status when it gets re-starting. If the `KafkaSourceDescriptor` is no longer available, the `ReadFromKafkaDoFn` will return `stop()`. Otherwise the DoFn will output records from this `KafkaSourceDescriptor` till next checkpoint happens.